azkustodata/kcsb.go

```diff
@@ -42,83 +46,118 @@ type ConnectionStringBuilder struct {
+	case keywords.InitialCatalog:
+		kcsb.InitialCatalog = value
+	case keywords.FederatedSecurity:
+		bval, err := strconv.ParseBool(value)
+		if err != nil {
+			return kustoErrors.ES(kustoErrors.OpUnknown, kustoErrors.KOther, "error: Couldn't parse federated security value: %s", err)
+		}
+		kcsb.AadFederatedSecurity = bval
+	case keywords.ApplicationClientId:
 		kcsb.ApplicationClientId = value
-	case applicationKey:
-		kcsb.ApplicationKey = value
-	case applicationCertificate:
-		kcsb.ApplicationCertificatePath = value
-	case sendCertificateChain:
-		bval, _ := strconv.ParseBool(value)
-		kcsb.SendCertificateChain = bval
-	case authorityId:
+	case keywords.UserId:
+		kcsb.AadUserID = value
+	case keywords.AuthorityId:
 		kcsb.AuthorityId = value
-	case applicationToken:
+
+	case keywords.ApplicationToken:
 		kcsb.ApplicationToken = value
-	case userToken:
+	case keywords.UserToken:
 		kcsb.UserToken = value
-	case interactiveLogin:
-		bval, _ := strconv.ParseBool(value)
-		kcsb.InteractiveLogin = bval
-	case domainHint:
-		kcsb.RedirectURL = value
+	case keywords.ApplicationKey:
+		kcsb.ApplicationKey = value
+	case keywords.ApplicationCertificateX5C:
+		bval, err := strconv.ParseBool(value)
+		if err != nil {
+			return kustoErrors.ES(kustoErrors.OpUnknown, kustoErrors.KOther, "error: Couldn't parse certificate x5c value: %s", err)
+		}
+		kcsb.SendCertificateChain = bval
+	case keywords.ApplicationCertificateBlob:
+		decodeString, err := base64.StdEncoding.DecodeString(value)
+		if err != nil {
+			return kustoErrors.ES(kustoErrors.OpUnknown, kustoErrors.KOther, "error: Couldn't decode certificate blob: %s", err)
+		}
+		kcsb.ApplicationCertificateBytes = decodeString
+	case keywords.ApplicationNameForTracing:
+		kcsb.ApplicationForTracing = value
+	case keywords.UserNameForTracing:
+		kcsb.UserForTracing = value
+	case keywords.Password:
+		kcsb.Password = value
 	}
+
 	return nil
 }
 
+// ConnectionString Generates a connection string from the current state of the ConnectionStringBuilder.
+func (kcsb *ConnectionStringBuilder) ConnectionString(secrets bool) (string, error) {
+	values := map[string]string{}
+
+	values[keywords.DataSource] = kcsb.DataSource
+	values[keywords.InitialCatalog] = kcsb.InitialCatalog
+	if kcsb.AadFederatedSecurity {
+		values[keywords.FederatedSecurity] = "true"
+	}
+	values[keywords.ApplicationClientId] = kcsb.ApplicationClientId
+	values[keywords.UserId] = kcsb.AadUserID
+	values[keywords.AuthorityId] = kcsb.AuthorityId
+	values[keywords.ApplicationToken] = kcsb.ApplicationToken
+	values[keywords.UserToken] = kcsb.UserToken
+	values[keywords.ApplicationKey] = kcsb.ApplicationKey
+	if kcsb.SendCertificateChain {
+		values[keywords.ApplicationCertificateX5C] = "true"
+	}
+
+	if !isEmpty(kcsb.ApplicationCertificatePath) {
+		values[keywords.ApplicationCertificateBlob] = base64.StdEncoding.EncodeToString(kcsb.ApplicationCertificateBytes)
+	}
+
+	builder := strings.Builder{}
+
+	for k, v := range values {
```

Comment:
this being a map I'm not sure how it will order things, but conventionally _datasource_ would be first

Reply:
Go is very annoying, I think that's the best solution I got